BitwardenShared/Core/Platform/Services/ConfigService.swift

```diff
@@ -112,62 +130,136 @@ class DefaultConfigService: ConfigService {
+        guard !isPreAuth else {
+            return await stateService.getPreAuthServerConfig()
+        }
+        return try? await stateService.getServerConfig()
+    }
+
+    func configPublisher() async throws -> AsyncThrowingPublisher<AnyPublisher<MetaServerConfig?, Never>> {
+        configSubject.eraseToAnyPublisher().values
+    }
+
+    /// Sets the server config in state depending on if the call is being done before authentication.
+    /// - Parameters:
+    ///   - config: Config to set
+    ///   - isPreAuth: If true, the call is coming before the user is auth or when adding a new account
+    ///   - userId: The userId to set the server config to.
+    func setStateServerConfig(_ config: ServerConfig, isPreAuth: Bool, userId: String? = nil) async throws {
+        guard !isPreAuth else {
+            await stateService.setPreAuthServerConfig(config: config)
+            return
+        }
+        try? await stateService.setServerConfig(config, userId: userId)
+    }
+
     /// Performs a call to the server to get the latest config and updates the local value.
-    private func updateConfigFromServer() async {
+    /// - Parameter isPreAuth: If true, the call is coming before the user is auth or when adding a new account
+    private func updateConfigFromServer(isPreAuth: Bool) async {
+        // The userId is needed here so we know which user trigger getting the config
+        // which helps if this is done in background and the user somehow changes the user
+        // while this is loading.
+        let userId = try? await stateService.getActiveAccountId()
+
         do {
             let configResponse = try await configApiService.getConfig()
             let serverConfig = ServerConfig(
                 date: timeProvider.presentTime,
                 responseModel: configResponse
             )
-            try? await stateService.setServerConfig(serverConfig)
+            try? await setStateServerConfig(serverConfig, isPreAuth: isPreAuth, userId: userId)
+
+            configSubject.send(MetaServerConfig(isPreAuth: isPreAuth, userId: userId, serverConfig: serverConfig))
         } catch {
             errorReporter.log(error: error)
+
+            guard !isPreAuth else {
+                return
+            }
+
+            let localConfig = try? await stateService.getServerConfig(userId: userId)
+            guard localConfig == nil,
+                  let preAuthConfig = await stateService.getPreAuthServerConfig() else {
+                return
+            }
+
+            try? await setStateServerConfig(preAuthConfig, isPreAuth: false, userId: userId)
         }
     }
 }
+
+/// Helper object to send updated server config object with extra metadata
+/// like whether it comes from pre-auth and the user ID it belongs to.
+/// This is useful for getting the config on background and establishing which was the original context.
+struct MetaServerConfig {
```

Comment:
🤔 I'm not sure I love this name particularly. Maybe something like `ServerConfigInfo`? Not sure I like that either. It's probably fine, but I wouldn't mind finding a better name.

Reply:
Yeah I struggled to find a name for it as well. Not a fan of `ServerConfigInfo`, what about `ServerConfigMessage`? As it gets inline with publishing a message and subscribing to a message sort of speak.

Reply:
Just for the sake of the upcoming release, can we leave `MetaServerConfig`? We can create a tech-debt task to change the name if you wish so we don't block the PR for this.